omicverse/Space/Space.py

```diff
@@ -0,0 +1,210 @@
+            adata_sp.var.index = [g.lower() for g in adata_sp.var.index]
+            genes = list(g.lower() for g in genes)
+
+        adata_sc.var_names_make_unique()
+        adata_sp.var_names_make_unique()
+    
+
+    # Refine `marker_genes` so that they are shared by both adatas
+        genes = list(set(genes) & set(adata_sc.var.index) & set(adata_sp.var.index))
+    # logging.info(f"{len(genes)} shared marker genes.")
+
+        adata_sc.uns["training_genes"] = genes
+        adata_sp.uns["training_genes"] = genes
+        logging.info(
+            "{} training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.".format(
+                len(genes)
+            )
+        )
+
+    # Find overlap genes between two AnnDatas
+        overlap_genes = list(set(adata_sc.var.index) & set(adata_sp.var.index))
+    # logging.info(f"{len(overlap_genes)} shared genes.")
+
+        adata_sc.uns["overlap_genes"] = overlap_genes
+        adata_sp.uns["overlap_genes"] = overlap_genes
+        logging.info(
+            "{} overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.".format(
+                len(overlap_genes)
+            )
+        )
+
+    # Calculate uniform density prior as 1/number_of_spots
+        adata_sp.obs["uniform_density"] = np.ones(adata_sp.X.shape[0]) / adata_sp.X.shape[0]
+        logging.info(f"uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata."
+        )
+
+    # Calculate rna_count_based density prior as % of rna molecule count
+        rna_count_per_spot = np.array(adata_sp.X.sum(axis=1)).squeeze()
+        adata_sp.obs["rna_count_based_density"] = rna_count_per_spot / np.sum(rna_count_per_spot)
+        logging.info(
+            f"rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata."
+        )
+ 
+    def adata_to_cluster_expression(adata, cluster_label, scale=True, add_density=True):
+
+        try:
+            value_counts = adata.obs[cluster_label].value_counts(normalize=True)
+        except KeyError as e:
+            raise ValueError("Provided label must belong to adata.obs.")
+        unique_labels = value_counts.index
+        new_obs = pd.DataFrame({cluster_label: unique_labels})
+        adata_ret = sc.AnnData(obs=new_obs, var=adata.var, uns=adata.uns)
+
+        X_new = np.empty((len(unique_labels), adata.shape[1]))
+        for index, l in enumerate(unique_labels):
+            if not scale:
+                X_new[index] = adata[adata.obs[cluster_label] == l].X.mean(axis=0)
+            else:
+                X_new[index] = adata[adata.obs[cluster_label] == l].X.sum(axis=0)
+        adata_ret.X = X_new
+
+        if add_density:
+            adata_ret.obs["cluster_density"] = adata_ret.obs[cluster_label].map(
+                lambda i: value_counts[i]
+            )
+
```

Comment:
You can use `self.adata_sc` and `self.adata_st` to store the data and `map_cells_to_space` don't need to input `adata_sc` and `adata_st`